### ISSUE TITLE
bug in arithmetic metaqueries

### ISSUE THREAD

**Issue by gwuillou** [GoshPosh/grafana-meta-queries#29]
I am doing a sum of 3 queries A + B +C . I get the follow result:

![issues_mq](https://user-images.githubusercontent.com/20662503/42308580-4ce44908-8036-11e8-96da-d5339df42235.png)

But if I zoom out, there is not anymore this 0 value in the left corner and everything works.

I am using influxdb and grafana 5.1.3.

**Comment by Gauravshah** [comment]
@gwuillou when zoomed-in  are the results correct ?
When zoomed out are able to see A,B,C ( without applying A+B+C Query) ?

**Comment by gwuillou** [comment]
Here different zooms. There is not always the problem apparently!

![issues_mq2](https://user-images.githubusercontent.com/20662503/42309328-8c24d7d4-8038-11e8-81ea-8bbc605cbf27.png)
![issues_mq3](https://user-images.githubusercontent.com/20662503/42309329-8c3bc0fc-8038-11e8-9c3d-bf095b801737.png)
![issues_mq4](https://user-images.githubusercontent.com/20662503/42309330-8c530776-8038-11e8-9472-8bf5fd9ebf12.png)
![issues_mq5](https://user-images.githubusercontent.com/20662503/42309331-8c692b1e-8038-11e8-92bd-80bb56bd3caf.png)
![issues_mq6](https://user-images.githubusercontent.com/20662503/42309332-8c7ff79a-8038-11e8-9155-8ac5ef8ba0f7.png)

**Comment by Gauravshah** [comment]
can you send a screenshot of your panel as well ( dashboard config )

**Comment by gwuillou** [comment]
[issues_MQ7.pdf](https://github.com/GoshPosh/grafana-meta-queries/files/2166130/issues_MQ7.pdf)

**Comment by Gauravshah** [comment]
I think the problem is because metaqueries does not assume null as 0. So for the sections where 1 of the (a,b or c ) value is null then the resultant D is null. You can probably try (((a)? a :0) + ((b)? b :0) + ((c)? c :0))

**Comment by gwuillou** [comment]
Hello,
I have tried your proposition without success: (((A['siesa'])?A['siesa']:0)+((C['djeva'])?C['djeva']:0)+((B['sinergy'])?B['sinergy']:0))
I don’t know if you are other idea?

Thank for your help,

Gaëtan

**Comment by Gauravshah** [comment]
I am not sure, I am unable to reproduce the error on my end. Looks to me that one of the datapoint is throwing the graph off. Can you try plotting a table instead and look at the odd data point ?

**Comment by gwuillou** [comment]
When I am exporting the csv:

Conso_FMV;2018-07-05T21:00:00+02:00;30.117048
Conso_FMV;2018-07-05T21:15:00+02:00;29.324952
Conso_FMV;2018-07-05T21:30:00+02:00;29.627704
Conso_FMV;2018-07-05T21:45:00+02:00;29.00208
Conso_FMV;2018-07-05T22:00:00+02:00;29.310443999999997
Conso_FMV;2018-07-05T22:15:00+02:00;29.481824
Conso_FMV;2018-07-05T22:30:00+02:00;28.184124
Conso_FMV;2018-07-05T22:45:00+02:00;28.682604
Conso_FMV;2018-07-05T23:00:00+02:00;28.003059999999998
Conso_FMV;2018-07-05T23:15:00+02:00;26.963816
Conso_FMV;2018-07-05T23:30:00+02:00;26.033584000000005
Conso_FMV;2018-07-05T23:45:00+02:00;26.090695999999998
Conso_FMV;2018-07-06T00:00:00+02:00;24.468472
Conso_FMV;2018-06-25T02:00:00+02:00;0
Conso_FMV;2018-07-06T00:15:00+02:00;0
Conso_FMV;2018-07-06T00:30:00+02:00;0
Conso_FMV;2018-07-06T00:45:00+02:00;0
Conso_FMV;2018-07-06T01:00:00+02:00;0
Conso_FMV;2018-07-06T01:15:00+02:00;0
Conso_FMV;2018-07-06T01:30:00+02:00;0
Conso_FMV;2018-07-06T01:45:00+02:00;0
Conso_FMV;2018-07-06T02:00:00+02:00;0
Conso_FMV;2018-07-06T02:15:00+02:00;0
Conso_FMV;2018-07-06T02:30:00+02:00;0

**Comment by Gauravshah** [comment]
can you give a try to `(((A['siesa'])?Math.round(A['siesa']*100)/100:0)+((C['djeva'])?Math.round(C['djeva']*100)/100:0)+((B['sinergy'])?Math.round(B['sinergy']*100)/100:0))`

**Comment by gwuillou** [comment]
Always the same problem!

**Comment by Gauravshah** [comment]
its hard to debug without that dataset

**Comment by gwuillou** [comment]
Perhaps it can help you:

**Comment by TheRealKerni** [comment]
I have similar drawing issues. I can reproduce it when showing 2 datasets with data available in different time intervals and selecting an 'group by time()' interval that is smaller than the first but larger than the second.

E.g. Dataset 1 is available in 10 Minutes interval, dataset 2 in 5 Minutes intervals. using 'group by time(10m)' shows nice graphs, but using 'group by time (5m)' shows issues in the graph. I am not arguing about the sense of this grouping, but to provide an example on how to reproduce graphing issues with MetaQueries.

Hope this helps.

**Comment by Gauravshah** [comment]
@TheRealKerni yup that scenario makes sense. but I am unsure if we can default numbers to 0. Its more general grafana issue than this plugin itself. I think it behaves in the same way if a datasource itself would have had no points in that location

**Comment by TheRealKerni** [comment]
I am not sure what is going on yet, tried to debug a bit but haven't find the cause.

But I have a small example animated gif that may help.

The data source has data that is fetched once a minute, the minimum interval for the chart is set to 5 minutes. When showing the influxDB data, everything is fine, when showing the MetaQuery instead it acts app. I found that the timestamp of the last item when using MetaQuery suddenly **changes** to the **first** timestamp in the dataset!

Take a look.

![example](https://user-images.githubusercontent.com/42185/52901100-cf3f5800-31fe-11e9-9954-9114837da4ef.gif)

**Comment by Gauravshah** [comment]
@ShilpaSivanesan we should see if this problem exists in a generic fashion. Lets look into this one next week.

**Comment by tiago-cruz-movile** [comment]
Hello,

I have two timestamps, and I would like to see the difference between using `B['dequeuedAt']-A['routedAt']`

Sometimes I've got strange negative values, like this:

![image](https://user-images.githubusercontent.com/4048174/62329961-bc0f0480-b48d-11e9-9e4b-1c06ec57136b.png)

How can I avoid null values?

I've tried 
`(((B['dequeuedAt'])? B['dequeuedAt'] :0) - ((A['routedAt'])? A['routedAt'] :0)`
and
`((B['dequeuedAt'])?Math.round(B['dequeuedAt']*100)/100:0) - ((A['routedAt'])?Math.round(A['routedAt']*100)/100:0)`

But without success.

Sometimes the count got really strange like this:

![image](https://user-images.githubusercontent.com/4048174/62330073-fb3d5580-b48d-11e9-9bf2-a422abd0b52a.png)

Thanks!!

**Comment by Gauravshah** [comment]
@tiago-cruz-movile you should put it on a table to see what results do you get.
also you can try `(  (B['dequeuedAt']  && A['routedAt'])? B['dequeuedAt'] - A['routedAt'] : 0 )`

**Comment by aquilax** [comment]
I had similar problem two queries `A` and `B` and the `B` query did not return values for every time interval for which `A` did. What worked for me was using something like `A['key1'] + (B ? B['key2'] : 0)` which essentially returns 0 if the B query does not have a value.

**Comment by ZyanKLee** [comment]
today I noticed a very similar behaviour with a simple A-B logic:

A: `SELECT mean("value") AS "Free" FROM "node_memory_MemFree_bytes" WHERE $timeFilter GROUP BY time($__interval) fill(none)`
B: `SELECT mean("value") AS "Total" FROM "node_memory_MemTotal_bytes" WHERE $timeFilter GROUP BY time($__interval) fill(none)`
C: `B["Total"]-A["Free"]`

At first I had the time() set to `10s` statically - and when I plotted 13 days or more the effect would look like that: 

![Bildschirmfoto 2020-02-03 um 17 20 18](https://user-images.githubusercontent.com/1178490/73670412-78e65600-46a9-11ea-80b0-dfe75bd751dd.png)

then I went back to letting grafana determine the grouping interval and it solved the problem for me. (tested up till 30 days worth of data)

**Comment by cassiel74** [comment]
Can't get metaqueries to work like you do...
Here's my setup, what I am missing?
![image](https://user-images.githubusercontent.com/6669940/75010379-42685380-547d-11ea-9f05-da841b522239.png)

Regards

**Comment by TheRealKerni** [comment]
@cassiel74 add the value `$col` into the big `ALIAS BY` field.

The problem is that if `ALIAS BY` is not set, then the actual identifier prepended by `values.`. The `$col` value will eliminate this prepending string.

**Comment by cassiel74** [comment]
Ah... I put the alias in SQL style :-)
Now it's working, using only `ALIAS BY` and nothing in SELECT statement

**Comment by Gauravshah** [comment]
thank you @TheRealKerni

**Comment by frams** [comment]
I was having the same issue I solve mine by grouping by time ($__interval, previous) on both A AN B queries (MYSQL Data Source) and that seems to have fixed the issue.
![image](https://user-images.githubusercontent.com/2160085/88823372-ff7c7b80-d192-11ea-9684-6bb4d91d3126.png)
